### PR TITLE
wrappers: install D-Bus service activation files for snapd session tools on core

### PR DESCRIPTION
**Pull request by jhenstridge**
At present, the D-Bus service activation files for userd and the session agent are not installed on Ubuntu Core 18+ systems.

They were available on Ubuntu Core 16 through the core snap, but on newer systems they're stuck in the snapd snap, where dbus-daemon won't look for them.  This PR extends the `wrappers/core18.go` code to install them to `/var/lib/snapd/dbus-1/services`, which was added to dbus-daemon's search path for the benefit of dbus activatable snap services.

The existing "snap userd" services used to back the xdg-open and xdg-settings proxies are not currently useful on Ubuntu Core (as they simply call the binaries in the host file system, which proxy back to userd on Core), so I've updated them to explicitly error out on these systems.  I've also added a spread test demonstrating this with the /usr/bin/xdg-open proxy.

### CODE REVIEW

**Comment by jhenstridge** [comment]
I think this is ready for review now.  The new spread test running on ubuntu-core-20-64 verifies that the service activation files are found by a session bus running on UC, even though all the current APIs are stubbed out.

File: wrappers/core18.go [hunk 1]
```diff
@@ -586,6 +617,9 @@ func RemoveSnapdSnapServicesOnCore(s *snap.Info, inter interacter) error {
 
 	sysd := systemd.NewUnderRoot(dirs.GlobalRootDir, systemd.SystemMode, inter)
 
+	if err := undoSnapdDbusActivationOnCore(s); err != nil {
```
Comment (pedronis) [inline wrappers/core18.go:620]: is this bit tested (I haven't looked how core18_test.go is structured) ?

File: wrappers/core18_test.go [hunk 2]
```diff
@@ -165,6 +169,15 @@ WantedBy=snapd.service
 	}, {
 		filepath.Join(dirs.SnapDBusSessionPolicyDir, "snapd.session-services.conf"),
 		"<busconfig/>",
+	}, {
```
Comment (pedronis) [inline wrappers/core18_test.go:172]: we might want also a new or some relevant preexisting spread core tests to check the files are present now

**Review by pedronis — COMMENTED** [review]
looks reasonable, some test questions

File: wrappers/core18.go (hunk 1 above)
Comment (jhenstridge) [reply wrappers/core18.go:620]: I've added to some of the other existing tests to show that the D-Bus config files are not copied on classic systems (where code in `interfaces/dbus/backend.go` currently tries to handle this), and that they are removed when the snapd snap is uninstalled.

**Review by pedronis — APPROVED** [review]
thank you

File: usersession/userd/launcher_test.go [hunk 3]
```diff
@@ -31,32 +31,32 @@ import (
...
 	s.launcher = &userd.Launcher{}
 	s.mockXdgOpen = testutil.MockCommand(c, "xdg-open", "")
-	s.restoreSnapFromSender = userd.MockSnapFromSender(func(*dbus.Conn, dbus.Sender) (string, error) {
+	s.AddCleanup(s.mockXdgOpen.Restore)
+	s.AddCleanup(userd.MockSnapFromSender(func(*dbus.Conn, dbus.Sender) (string, error) {
 		return "some-snap", nil
-	})
-}
-
-func (s *launcherSuite) TearDownTest(c *C) {
```
Comment (anonymouse64) [inline usersession/userd/launcher_test.go:57]: thanks for this cleanup

File: tests/main/xdg-open-on-core/task.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+summary: Ensure snap userd returns an error on Ubuntu Core
+
+systems:
+    - ubuntu-core-*
```
Comment (anonymouse64) [inline tests/main/xdg-open-on-core/task.yaml:4]: if this test is meant to be run on only core devices, you could move it to the `tests/core/...` suite instead where only ubuntu core systems are available to run

**Review by anonymouse64 — APPROVED** [review]
lgtm, but one of the spread tests in `tests/main` can move to `tests/core` I think

**Pull request merged**